### PR TITLE
End-To-End Automation: Inject data using OpenMrs Apis

### PR DESCRIPTION
**Pull request by drryanjames**
The major feature added in this pull request is the ability to communicate with the OpenMrs REST APIs to inject data into the database in preparation for an e2e test. Look at the second test in login_test.js for an example of how the Apis are used in a test, and in openMrsApis.js for a definition of the classes encapsulating API requests.

I also added the end to end tests under the test folder. Unit tests are in /test/unitTests and e2e tests are in /test/e2eTests. I think it's a bit more organized. Ideally, these changes would be in a separate pull request. I included them in this one because of the order of my commits. If this bothers anyone I can break things up.

### CODE REVIEW

**Comment by ynurmahomed** [comment]
It looks good but `Attempt to login with a user that does not have the provider role` is failing for me even though the user is created. @edrisse is it working for you?

```
create user request failed {
  "statusCode": 500,
  "body": {
    "error": {
      "message": "[null]",
      "code": "org.openmrs.api.context.UserContext:428",
      "detail": "java.lang.NullPointerException\n\tat org.openmrs.api.context.UserContext.setUserLocation(UserContext.java:428)\n\tat org.openmrs.api.context.UserContext.refreshAuthenticatedUser(UserContext.java:124)\n\tat org.openmrs.api.context.Context.refreshAuthenticatedUser(Context.java:326)\n\tat org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.UserResource1_8.save(UserResource1_8.java:146)\n\tat org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.UserResource1_8.save(UserResource1_8.java:46)\n\tat org.openmrs.module.webservices.rest.web.resource.impl.DelegatingCrudResource.create(DelegatingCrudResource.java:78)\n\tat org.openmrs.module.webservices.rest.web.v1_0.controller.MainResourceController.create(MainResourceController.java:92)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:606)\n\tat org.springframework.web.bind.annotation.support.HandlerMethodInvoker.invokeHandlerMethod(HandlerMethodInvoker.java:176)\n\tat org.springframework.web.servlet.mvc.annotation.AnnotationMethodHandlerAdapter.invokeHandlerMethod(AnnotationMethodHandlerAdapter.java:440)\n\tat org.springframework.web.servlet.mvc.annotation.AnnotationMethodHandlerAdapter.handle(AnnotationMethodHandlerAdapter.java:428)\n\tat org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:925)\n\tat org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:856)\n\tat org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:953)\n\tat org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:855)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:650)\n\tat org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:829)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:731)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.module.web.filter.ForcePasswordChangeFilter.doFilter(ForcePasswordChangeFilter.java:61)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.module.web.filter.ModuleFilterChain.doFilter(ModuleFilterChain.java:72)\n\tat org.openmrs.web.filter.GZIPFilter.doFilterInternal(GZIPFilter.java:64)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)\n\tat org.openmrs.module.web.filter.ModuleFilterChain.doFilter(ModuleFilterChain.java:70)\n\tat org.openmrs.module.webservices.rest.web.filter.AuthorizationFilter.doFilter(AuthorizationFilter.java:104)\n\tat org.openmrs.module.web.filter.ModuleFilterChain.doFilter(ModuleFilterChain.java:70)\n\tat org.springframework.web.filter.ShallowEtagHeaderFilter.doFilterInternal(ShallowEtagHeaderFilter.java:73)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)\n\tat org.openmrs.module.web.filter.ModuleFilterChain.doFilter(ModuleFilterChain.java:70)\n\tat org.openmrs.module.web.filter.ModuleFilter.doFilter(ModuleFilter.java:54)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.web.filter.OpenmrsFilter.doFilterInternal(OpenmrsFilter.java:109)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.springframework.orm.hibernate3.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:230)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:105)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)\n\tat org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)\n\tat org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)\n\tat org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)\n\tat org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)\n\tat org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:110)\n\tat org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:498)\n\tat org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)\n\tat org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)\n\tat org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:1025)\n\tat org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)\n\tat org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:445)\n\tat org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1115)\n\tat org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:637)\n\tat org.apache.tomcat.util.net.AprEndpoint$SocketWithOptionsProcessor.run(AprEndpoint.java:2492)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:748)\n"
    }
  },
  "headers": {
    "server": "Apache-Coyote/1.1",
    "set-cookie": [
      "JSESSIONID=DDBC74536F181D9ACBA050086499F79E; Path=/openmrs"
    ],
    "content-type": "application/json;charset=UTF-8",
    "content-length": "7409",
    "date": "Mon, 07 May 2018 08:22:56 GMT",
    "connection": "close"
  },
  "request": {
    "uri": {
      "protocol": "http:",
      "slashes": true,
      "auth": null,
      "host": "localhost:8080",
      "port": "8080",
      "hostname": "localhost",
      "hash": null,
      "search": null,
      "query": null,
      "pathname": "/openmrs/ws/rest/v1/user",
      "path": "/openmrs/ws/rest/v1/user",
      "href": "http://localhost:8080/openmrs/ws/rest/v1/user"
    },
    "method": "post",
    "headers": {
      "Content-Type": "application/json",
      "Authorization": "Basic YWRtaW46ZVNhdWRlMTIz",
      "content-length": 154
    }
  }
}
```

File: test/e2eTests/tests/login_test.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	// Log the user out
+	const logoutStatus = dashboardPage.logout()
+	
+	// Validate that logout was successful
+	logoutStatus.successful()
+})
+
+Scenario('Attempt to login with a user that does not have the provider role', async (I, Apis, LoginPage, Data) => {
+	// Create a user that does not have the provider role
+	await Apis.user.create(Data.users.nonProvider)
```
Comment (ynurmahomed) [inline test/e2eTests/tests/login_test.js]: Should we remove the user after the test runs?

**Comment by edrisse** [comment]
@drryanjames  and @ynurmahomed we can go ahead with this solution for now, but we will be better served having a back-end operation that performs a complete cleanup of the database. That way we don't have to maintain deletion of every record we create and no risk of compromising the result of following tests if the deletes of current test fails.

File: test/e2eTests/hooks.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+const event = require('codeceptjs').event;
+const Apis = require('./rest/openMrsApis')
+
+module.exports = function() {
+	// Fired after every test
+  event.dispatcher.on(event.test.after, function (test) {
+  	// Remove any and all data that was injected into the DB
+  	// after each test
+  	Apis.cleanUp()
```
Comment (drryanjames) [inline test/e2eTests/hooks.js]: @ynurmahomed @edrisse this line removes all data that was injected during a test. It allows tests to inject data without worrying about cleanup.

File: test/e2eTests/tests/login_test.js (hunk 1 above)
Comment (drryanjames) [reply test/e2eTests/tests/login_test.js]: @ynurmahomed @edrisse take a look at hooks.js. After every test all data that was injected is automatically removed so tests don't have to worry about doing it themselves.

File: test/e2eTests/hooks.js (hunk 2 above)
Comment (edrisse) [reply test/e2eTests/hooks.js]: @drryanjames ok, thx for explaining that. @ynurmahomed from my side i think we can merge.

**Comment by drryanjames** [comment]
Thanks guys

**Pull request merged**